modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module contains the implementation of the List type.
+
+  A List is a lightweight container similar to an array that is suitable for
+  building up and iterating over a collection of elements in a structured
+  manner.
+
+  The highly parallel nature of Chapel means that great care should be taken
+  when performing operations that may invalidate references to List elements.
+  Inserts and removals into the middle of a List are example operations that
+  may invalidate references. Appending an element to the end of a List will
+  never invalidate references to elements contained in the List.
+
+  The following operations may invalidate references to elements contained in
+  a List:
+
+      - insert
+      - remove
+      - sort
+      - pop
+      - clear
+
+  Additionally, all references to List elements are invalidated when the List
+  is deinitialized.
+
+  List objects are not parallel safe by default, but can be made parallel safe
+  by setting the param formal `parSafe` to true in any List constructor. When
+  constructed from another List, a List object will inherit the parallel
+  safety mode of its originating List.
+
+  Inserts and removals into a List are O(n) worst case and should be performed
+  with care. Access into a List object is currently O(log n), but is expected
+  to reduce to O(1) in the future.
+*/
+module Lists {
+
+  pragma "no doc"
+  config const _initialCapacity = 8;
+
+  pragma "no doc"
+  config const _initialBlockCapacity = 16;
+
+  pragma "no doc"
+  config param _sanityChecks = true;
+
+  //
+  // Some asserts are useful while developing, but can be turned off when the
+  // implementation is correct.
+  //
+  private inline proc _sanity(expr: bool) {
+    if _sanityChecks then
+      assert(expr);
+  }
+
+  //
+  // Basically just a wrapper around _ddata at this point, with one based
+  // indexing to avoid logical errors.
+  //
+  pragma "no doc"
+  class ListBlock {
```

Comment:
I'm pretty sure you should make `ListBlock` a record. Additionally, I'm pretty sure you should make `ListBlock` support 0-based indexing. I think that will simplify the other code (and you can subtract one once and for all in list.this / other index locations). You are already subtracting 1 to get the block index computed.